packages/ozone-helper/ozone-type/ozone/model/FlowrWarningLog.ts

```diff
@@ -0,0 +1,67 @@
+import {Instant, Item, OzoneType, UUID} from './Item'
+import {isLocationBuilding, LocationBuilding} from './LocationBuilding'
+import {PersistedFlowrWarningOperator} from './FlowrWarningOperator'
+import {isLocationSite, LocationSite} from './LocationSite'
+
+export enum LOG_ACTION {
+	LOGIN = 'LOGIN',
+	LOGOUT = 'LOGOUT',
+	SWITCH_SITE = 'SWITCH_SITE',
+	WARNING_ON = 'WARNING_ON',
+	WARNING_OFF = 'WARNING_OFF',
+	WARNING_TEST = 'WARNING_TEST'
+}
+
+class BareFlowrWarningLog extends Item {
+	date: Instant
+	action: LOG_ACTION
+	targetType?: string
+
+	constructor(src: BareFlowrWarningLog) {
+		super(src)
+		this.date = src.date
+		this.action = src.action
+		this.targetType = src.targetType
+	}
+}
+
+@OzoneType('flowr.warning.log')
+export class FlowrWarningLog extends BareFlowrWarningLog {
+	operator: UUID
+	target?: UUID
+
+	constructor(src: FlowrWarningLog) {
+		super(src)
+		this.operator = src.operator
+		this.target = src.target
+	}
+}
+
+export class HydratedFlowrWarningLog extends BareFlowrWarningLog {
+	id: UUID
+	operator: PersistedFlowrWarningOperator
+	target?: LocationBuilding | LocationSite
+
+	constructor(src: any) {
+		super(src)
+		this.id = src.id
+		this.operator = src.operator
+		this.target = src.target
+	}
+
+	getTargetToString() {
+		if (!this.target) return
+		if (isLocationSite(this.target)) return this.target.name
+		if (isLocationBuilding(this.target)) return this.target.warningButtonText
+	}
+}
+
+export const isFlowrWarningLog = (object: any): object is FlowrWarningLog => {
+	return object.type === 'flowr.warning.log'
+}
+
+export const isHydratedFlowrWarningLog = (object: any): object is HydratedFlowrWarningLog => {
+	return isFlowrWarningLog(object) &&
+		(object.target === undefined || typeof (object.target) === 'object') &&
```

Comment:
**instanceof** does not work without a bit of tinkering because **object.target** is a string in **FlowrWarningLog** and the **isFlowrWarningLog(object)** asserts that **object** is a **FlowrWarningLog**. There are typing errors in most IDEs.

Honestly, the only goal is to check that **object** is a **FlowrWarningLog** whose property **target** is an object instead of a string (an id). I don't feel like we gain much in restricting more while decreasing readability.

You might see a simple solution that I don't see, though.